### ISSUE TITLE
False positive iterating over and ignoring 0..n + 1

### ISSUE THREAD

**Issue by TomFryersMidsummer** [rust-lang/rust-clippy#11719]
### Summary

The idiomatic way to do the same thing `k` times is
```rust
for _ in 0..k {
    do_thing();
}
```
Suppose `k` is the expression `n + 1`.
```rust
for _ in 0..n + 1 {
    do_thing();
}
```
This now triggers the `range_plus_one` lint, but
```rust
for _ in 0..=n {
    do_thing();
}
```
is much less clear about what's going on.

`(0..n + 1).map(|_| f())` is similar.

### Lint Name

range_plus_one

### Reproducer

I tried this code:

```rust
fn foo(n: usize) {
    for _ in 0..n + 1 {
        println!("foo");
    }
}
```

with the pedantic group enabled.

I saw this happen:

```
warning: an inclusive range would be more readable
```

I expected to see this happen:

[no warning]

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-unknown-linux-gnu
release: 1.73.0
LLVM version: 17.0.2
```

**Comment by llogiq** [comment]
I don't agree that "it's much less clear" – that's what an inclusive range is for.

Could you please elaborate on your reasoning?

**Comment by TomFryersMidsummer** [comment]
In my mind, an inclusive range is for performing some action for each of the numbers `x..=y`, such as
```rust
for i in 0..=5 {
    print!("{i}");
}
```
gives `012345`.

The main reason to prefer inclusive range over an exclusive ones is that it's often intuitive to
have `for i in x..=y {}` include a loop iteration in which `i` is `y`.

However, in this example, we're not doing anything with the numbers, so that never happens.

```rust
for _ in 0..n + 1 {
    do_thing();
}
```
and
```rust
for _ in 0..=n {
    do_thing();
}
```

are both expressions of the pseudo-code

```
repeat n + 1 times:
    do_thing();
```

Unless the reason you're doing something `n` + 1 times is specifically _because_ you want to do
it once for each of the numbers `0..=n`, it seems a bit convoluted to use the fact that
`0..=n` happens to have `n` + 1 elements as a way to repeat something `n` + 1 times.

Put another way, without this exception:
```rust
// Repeat something 5 times
for _ in 0..5 {}
// Repeat something k times
for _ in 0..k {}
// Repeat something 2 * k times
for _ in 0..2 * k {}
// Repeat something k + 1 times
for _ in 0..=k {} // ?
```

**Comment by llogiq** [comment]
> In my mind, an inclusive range is for performing some action for each of the numbers x..=y, ...

In *my* mind (and as per their documentation), inclusive ranges are for including the upper bound. You'll note that this has nothing to do with actions or numbers. What's more, the `y + 1` might overflow, which could lead to the loop being skipped completely, so iterating over `x ..= y` should be seen as a safe default.

With that said, there may be a performance angle that benefits the exclusive range, but I'd like to see benchmarks before jumping to conclusions.

**Comment by ronnodas** [comment]
I agree that the semantics of `0..n+1` *is* preferable to `0..=n` *in this context*. Have you considered `_ in 1 ..= k + 1` instead (and similarly `1..=k`, `1..=2 * k` etc)?

**Comment by TomFryersMidsummer** [comment]
> I agree that the semantics of `0..k+1` _is_ preferable to `0..=k` _in this context_. Have you considered `_ in 1..=k + 1` instead (and similarly `1..=k`, `1..=2 * k` etc)?

Using `1..=expr` everywhere instead of `0..expr` does fix this problem, but I don't think that method is ideal, since Rust generally counts things from zero, and tends to prefer `..` to `..=` ranges. Although that's not a huge deal here, since the numbers aren't actually being used.

**Comment by TomFryersMidsummer** [comment]
> In _my_ mind (and as per their documentation), inclusive ranges are for including the upper bound.

Literally, yes, and if you just look at the expression `0..k + 1`, without any surrounding context, I agree that `0..=k` looks preferable. However, in this context, it seems to me that it's just making the code more difficult to understand.
